model/task_queue_service_dependency.go

```diff
@@ -239,8 +238,6 @@ func (d *basicCachedDAGDispatcherImpl) rebuild(items []TaskQueueItem) error {
 
 // FindNextTask returns the next dispatchable task in the queue, and returns the tasks that need to be checked for dependencies.
 func (d *basicCachedDAGDispatcherImpl) FindNextTask(ctx context.Context, spec TaskSpec, amiUpdatedTime time.Time) *TaskQueueItem {
-	d.mu.Lock()
```

Comment:
Removing the lock creates a race to access `d.sorted` and the items within it, so I suspect more bugs would start appearing related to dispatching if it was removed.

Also, can we write a unit test to verify that this is thread-safe? The race detector will identify if it can race to access `d.sorted` or any of its items.

Reply:
I refactored some of the functions associated with reading / writing the queue state into helper functions that use appropriate locks. I wrote a couple tests affirming that the function is thread safe for both regular and task group tasks and confirmed they also pass in the [race detector](https://spruce.mongodb.com/version/66e28ee248bbb600073711fc/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC)